File: FileThumbnails/js/js/scenario4.js
//// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
//// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
//// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//// PARTICULAR PURPOSE.
////
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    var page = WinJS.UI.Pages.define("/html/scenario4.html", {
        ready: function (element, options) {
            document.getElementById("folder-thumb-start").addEventListener("click", pickFolder, false);
        }
    });

    function pickFolder() {

        // Set API parameters
        var requestedSize = 200,
            thumbnailMode = Windows.Storage.FileProperties.ThumbnailMode.picturesView;

        // Clean output in case of repeat usage
        cleanOutput();

        var openpicker = new Windows.Storage.Pickers.FolderPicker;
        openpicker.fileTypeFilter.replaceAll([".jpg", ".png", ".bmp", ".gif", ".tif"]);
        openpicker.suggestedStartLocation = Windows.Storage.Pickers.PickerLocationId.picturesLibrary;
        openpicker.pickSingleFolderAsync().done(function (folder) {
            if (folder) {
                folder.getThumbnailAsync(thumbnailMode, requestedSize).done(
                    function (thumbnail) {
                        if (thumbnail) {
                            outputResult(folder, thumbnail, "ThumbnailMode.picturesView", requestedSize);
                        } else {
                            WinJS.log && WinJS.log(SdkSample.errors.noImages, "sample", "status");
                        }
                    },
                    function (error) {
                        WinJS.log && WinJS.log(SdkSample.errors.fail, "sample", "status");
                    }
                );
            } else {
                WinJS.log && WinJS.log(SdkSample.errors.cancel, "sample", "status");
            }
        });
    }

    function outputResult(item, thumbnailImage, thumbnailMode, requestedSize) {
        document.getElementById("folder-thumb-imageHolder").src = URL.createObjectURL(thumbnailImage, { oneTimeOnly: true });
        // Close the thumbnail stream once the image is loaded
        document.getElementById("folder-thumb-imageHolder").onload = function () {
            thumbnailImage.close();
        };
        document.getElementById("folder-thumb-modeName").innerText = thumbnailMode;
        document.getElementById("folder-thumb-folderName").innerText = "Folder used: " + item.name;
        document.getElementById("folder-thumb-requestedSize").innerText = "Requested size: " + requestedSize;
        document.getElementById("folder-thumb-returnedSize").innerText = "Returned size: " + thumbnailImage.originalWidth + "x" + thumbnailImage.originalHeight;
    }

    function cleanOutput() {
        WinJS.log && WinJS.log("", "sample", "status");
        document.getElementById("folder-thumb-imageHolder").src = "/images/placeholder-sdk.png";
        document.getElementById("folder-thumb-modeName").innerText = "";
        document.getElementById("folder-thumb-folderName").innerText = "";
        document.getElementById("folder-thumb-requestedSize").innerText = "";
        document.getElementById("folder-thumb-returnedSize").innerText = "";
    }
})();

File: Samples/FileThumbnails/cppwinrt/Scenario4.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario4.h"
#include "Scenario4.g.cpp"
#include "SampleConfiguration.h"

using namespace winrt;
using namespace Windows::Foundation;
using namespace Windows::Storage;
using namespace Windows::Storage::FileProperties;
using namespace Windows::Storage::Pickers;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Navigation;

namespace winrt::SDKTemplate::implementation
{
    Scenario4::Scenario4()
    {
        InitializeComponent();
    }

    void Scenario4::OnNavigatedTo(NavigationEventArgs const&)
    {
        Helpers::ResetOutput(rootPage, ThumbnailImage(), OutputTextBlock());
    }

    fire_and_forget Scenario4::GetThumbnailButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        Helpers::ResetOutput(rootPage, ThumbnailImage(), OutputTextBlock());

        // Pick a folder
        FolderPicker folderPicker;
        folderPicker.FileTypeFilter().ReplaceAll(FileExtensions::Image());
        StorageFolder folder = co_await folderPicker.PickSingleFolderAsync();

        if (folder != nullptr)
        {
            const ThumbnailMode thumbnailMode = ThumbnailMode::PicturesView;
            const uint32_t size = 200;
            StorageItemThumbnail thumbnail = co_await folder.GetThumbnailAsync(thumbnailMode, size);
            if (thumbnail != nullptr)
            {
                Helpers::DisplayResult(ThumbnailImage(), OutputTextBlock(), thumbnailMode, size, folder, thumbnail, false);
                thumbnail.Close();
            }
            else
            {
                rootPage.NotifyUser(Errors::NoIcon, NotifyType::StatusMessage);
            }
        }
        else
        {
            rootPage.NotifyUser(Errors::Cancel, NotifyType::StatusMessage);
        }
    }
}
